control-plane/agents/ha-node/src/detector.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    path_provider::{CachedNvmePathProvider, NvmePathNameCollection},
+    reporter::PathReporter,
+    Cli,
+};
+use futures::future::FutureExt;
+use nvmeadm::nvmf_subsystem::Subsystem;
+use std::{collections::HashMap, sync::Arc};
+use tokio::time::{sleep, Duration};
+
+/// Possible states of every path record.
+#[derive(Debug, Clone)]
+enum PathState {
+    Good,
+    Suspected,
+    Failed,
+}
+
+/// Object that represents a broken/suspected NVMe path.
+#[derive(Debug, Clone)]
+struct PathRecord {
+    /// Used to allow detection of outdated records.
+    epoch: u64,
+    nqn: String,
+    state: PathState,
+    reporter: Arc<PathReporter>,
```

Comment:
Why Arc? If it is supposed to be shared across thread, shouldn't PathReporter be protected by Mutex?

Reply:
Explained in my comment below.